platform/linux-generic/odp_crypto.c

```diff
@@ -308,6 +318,182 @@ void packet_hmac(odp_packet_t pkt,
 	HMAC_Final(ctx, hash, NULL);
 }
 
+static void do_pad_xor(uint8_t *out, const uint8_t *in, int len)
+{
+	int pos = 0;
+
+	for (pos = 0; pos < AES_BLOCK_SIZE; pos++, in++, out++) {
+		if (pos < len)
+			*out ^= *in;
+		if (pos >= len) {
+			*out ^= 0x80;
+			break;
+		}
+	}
+}
+
+static void xor_block(uint8_t *res, const uint8_t *op)
+{
+	int i;
+
+	for (i = 0; i < AES_BLOCK_SIZE; i++)
+		res[i] ^= op[i];
+}
+
+static
+void packet_aes_xcbc_mac(odp_packet_t pkt,
+			 const odp_crypto_packet_op_param_t *param,
+			 odp_crypto_generic_session_t *session,
+			 uint8_t *hash)
+{
+	uint8_t e[AES_BLOCK_SIZE] = {0};
+	uint8_t seg[AES_BLOCK_SIZE] = {0};
+	bool is_seg = false;
+	uint32_t seg_datalen = 0;
+	uint32_t offset = param->auth_range.offset;
+	uint32_t len   = param->auth_range.length;
+	uint32_t seglen = 0;
+	uint32_t datalen = 0;
+	int dummy_len = 0;
+	EVP_CIPHER_CTX *ctx;
+	void *mapaddr;
+	uint8_t *data = NULL;
+
+	ODP_ASSERT(offset + len <= odp_packet_len(pkt));
+	ODP_ASSERT(session != NULL);
+	ODP_ASSERT(sizeof(session->auth.key) >= 3 * AES_KEY_LENGTH);
+
+	ctx = EVP_CIPHER_CTX_new();
+	EVP_EncryptInit_ex(ctx, session->auth.evp_cipher,
+			   NULL, session->auth.key, NULL);
+
+	while (len > 0) {
+		mapaddr = odp_packet_offset(pkt, offset, &seglen, NULL);
+		datalen = seglen >= len ? len : seglen;
+		data = (uint8_t *)mapaddr;
+		offset += datalen;
+		len -= datalen;
+		/* Segmentation handle */
+		if (is_seg) {
+			is_seg = false;
+			if (datalen > AES_BLOCK_SIZE - seg_datalen) {
+				memcpy(seg + seg_datalen,
+				       data, AES_BLOCK_SIZE - seg_datalen);
+				xor_block(e, seg);
+				EVP_EncryptUpdate(ctx,
+						  e, &dummy_len, e, sizeof(e));
+				datalen -= AES_BLOCK_SIZE - seg_datalen;
+			} else {
+				memcpy(seg + seg_datalen, data, datalen);
+				data = seg;
+				datalen += seg_datalen;
+			}
+		}
+		for (; datalen > AES_BLOCK_SIZE; datalen -= AES_BLOCK_SIZE) {
+			xor_block(e, data);
+			EVP_EncryptUpdate(ctx, e, &dummy_len, e, sizeof(e));
+			data += AES_BLOCK_SIZE;
+		}
+		/* Segmentation handle */
+		if (datalen > 0 && len > 0) {
+			is_seg = true;
+			memcpy(seg, data, datalen);
+			seg_datalen = datalen;
+		}
+	}
+
+	if (data != NULL)
+		do_pad_xor(e, data, datalen);
```

Comment:
Pad should always be applied.

Reply:
Here is just a NULL pointer judgement. If the packet length is zero, it must be an error and we should not crash here.

Reply:
What is the mac of empty packet?

Reply:
ok, I can remove it.